pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/merge_rollup/MergeRollupTaskGenerator.java

```diff
@@ -0,0 +1,450 @@
+          })));
+
+      if (preSelectedSegmentsForOfflineTable.isEmpty()) {
+        LOGGER.warn("Skip generating task: {} for table: {}, no segment is found to merge.", taskType,
+            offlineTableName);
+        continue;
+      }
+
+      // Sort merge levels based on bucket time period
+      Map<String, Map<String, String>> mergeLevelToConfigs = MergeRollupTaskUtils.getLevelToConfigMap(taskConfigs);
+      List<Map.Entry<String, Map<String, String>>> sortedMergeLevelConfigs = mergeLevelToConfigs.entrySet()
+          .stream()
+          .sorted(Comparator.comparingLong(
+              e -> TimeUtils.convertPeriodToMillis(e.getValue().get(MinionConstants.MergeTask.BUCKET_TIME_PERIOD_KEY))))
+          .collect(Collectors.toList());
+
+      // Schedule tasks from lowest to highest merge level
+      String lowerMergeLevel = null;
+      long lowerMergeLevelWatermarkMs = -1;
+
+      for (Map.Entry<String, Map<String, String>> entry : sortedMergeLevelConfigs) {
+        String mergeLevel = entry.getKey();
+        Map<String, String> mergeConfig = entry.getValue();
+
+        // Get the bucket size and buffer
+        long bucketMs =
+            TimeUtils.convertPeriodToMillis(mergeConfig.get(MinionConstants.MergeTask.BUCKET_TIME_PERIOD_KEY));
+        long bufferMs =
+            TimeUtils.convertPeriodToMillis(mergeConfig.get(MinionConstants.MergeTask.BUFFER_TIME_PERIOD_KEY));
+
+        // Get watermark from MergeRollupTaskMetadata ZNode.
+        // windowStartMs = watermarkMs, windowEndMs = windowStartMs + bucketTimeMs
+        long waterMarkMs =
+            getWatermarkMs(offlineTableName, preSelectedSegmentsForOfflineTable.first().getStartTimeMs(), bucketMs,
+                mergeLevel);
+        long windowStartMs = waterMarkMs;
+        long windowEndMs = windowStartMs + bucketMs;
+
+        // Skip scheduling if there's incomplete task for current mergeLevel
+        if (inCompleteMergeLevelToTaskConfig.containsKey(mergeLevel)) {
+          LOGGER.info(
+              "Found incomplete tasks: {} of merge level: {} for the same table: {}, Skipping task generation: {}",
+              inCompleteMergeLevelToTaskConfig.get(mergeLevel), mergeLevel, offlineTableName, taskType);
+          lowerMergeLevel = mergeLevel;
+          lowerMergeLevelWatermarkMs = waterMarkMs;
+          continue;
+        }
+
+        if (!isValidMergeWindowEndTime(windowEndMs, bufferMs, lowerMergeLevel, lowerMergeLevelWatermarkMs)) {
+          LOGGER.warn(
+              "Window with start: {} and end: {} of mergeLevel: {} is not a valid merge window, Skipping task generation: {}",
+              windowStartMs, windowEndMs, mergeLevel, taskType);
+          lowerMergeLevel = mergeLevel;
+          lowerMergeLevelWatermarkMs = waterMarkMs;
+          continue;
+        }
+
+        // Find all segments overlapping with the merge window, if all overlapping segments are merged, bump up the target window
```

Comment:
Let's split it into 2 part, first pass bump up the time window, second pass gather the segments